### PR TITLE
Add new `dataset_predicate` param to `dc.load` for more flexible temporal filtering

### PR DESCRIPTION
**Pull request by robbibt**
### Reason for this pull request
The ability to query data from discrete, non-contiguous time ranges (e.g. for specific months or seasons over multiple years) has been a frequently requested ODC feature (e.g. see #868 and [here](https://gis.stackexchange.com/questions/402115/temporal-subsetting-of-landsat-collection-using-select-months-using-open-data-cu/402116#402116), plus multiple undocumented Slack queries). This can be extremely useful for analysing specific environmental conditions consistently over time, e.g. dry seasons in northern Australia, or Spring growing seasons in agricultural regions.

Previously, the only way this could be achieved is to use the `dc.find_datasets()` method to list available datasets, then remove items from this list before loading them via the `datasets` param in `dc.load`. This adds significant overhead to simple analyses, and can be overwhelming to new users who have only just mastered `dc.load` itself.

Recently, the `dea-tools` `load_ard` function and ODC Virtual Products have introduced a "predicate" param that allows users to easily filter data prior to loading. This would be powerful functionality to have in `dc.load` as well.

### Proposed changes
This PR adds a new `dataset_predicate` parameter to `dc.load`. This allows users to filter data prior to load using a simple predicate function. For example, the user could filter to only January observations across a multi-year period using the following syntax:

```
query = {
    "x": (153.45, 153.47),
    "y": (-28.90, -28.92),
    "time": ("2018", "2020"),
    "output_crs": "EPSG:3577",
    "resolution": (-10, 10),
    "group_by": "solar_day",
}

# Predicate function that returns true if month is Jan
def filter_jan(dataset):
    return dataset.time.begin.month == 1

# Specify using `dataset_predicate` param
ds = dc.load(
    product="s2a_ard_granule",
    measurements="nbart_red",
    dataset_predicate=filter_jan,
    dask_chunks={},
    **query
)
```
image

 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by alexgleith** [comment]
Is there a way we can test this?

**Comment by robbibt** [comment]
I don't have any experience writing tests but I'd be happy to learn so if someone could nudge me in the right direction!

**Comment by robbibt** [comment]
I don't have any experience writing tests but I'd be happy to learn so if someone could nudge me in the right direction!

For the above January example, this would verify the predicate was applied successfully: 
```
assert (ds.time.dt.month == 1).all()
```

**Comment by uchchwhash** [comment]
I think if I approve then it would automatically merge. So even though I approve of this very well-written contribution and am happy for this to be merged in, let me document my take on this.

It is unfortunate that `Datacube.load` is the the first point of contact with loading data for many users. I actually like `load_ard` more for this purpose, and am very pleased that people use it. As `load_ard` already performs this perfectly, I do not really see the value of this being added to the already cluttered API.

Perhaps more importantly, the long-standing Issue #542 means that quite possibly the addition of this to `Datacube.load` will convey a false sense of efficiency. The list of datasets will actually be loaded into memory before it is filtered. This is still more efficient than loading all the data and _then_ discarding some, but `load_ard` manages to do this anyway.

TL;DR `load_ard` already does this, and I like that

**Comment by robbibt** [comment]
Thanks @uchchwhash! This is my alternative perspective: `load_ard` is a great convenience function with some handy advanced features like predicate filtering. However, it's _very_ platform specific and designed specifically for Digital Earth Australia data. To take advantage of those advanced features, a user would need to make significant changes to the function (@cbur24 can testify to how tricky it was to adapt that func to DE Africa's data offerings). This would require significant knowledge of ODC and Python, which is can be a major blocker to many new and less code-proficient users.

On the other hand, the ability to apply advanced filtering on datasets prior to load is fundamental functionality that I think would benefit the ODC community in general, not just the subset of Digital Earth Australia users. I think there might also be a bit of a disconnect here about the role of `Datacube.load`: for many of our non-developer users, `Datacube.load` _is_ Open Data Cube. If I was to guess, `dc.load` probably makes up 90% of calls to ODC funcs in user code, so is by far the main way most people interact with the datacube. As such, I think there's a lot of value in continuing to refine this function so that it's as useful as it can be for common spatiotemporal processing tasks. 

The performance issue you raise is very relevant though - my feeling is however that this is a broader issue inherent to `dc.load` - at this stage the PR above describes `dataset_predicate` primarily as functionality to do useful things we can't do (e.g. non-continuous temporal load), not for performance reasons.

**Comment by uchchwhash** [comment]
😭 understood. If that's the reality then I suppose we will just have to live with `Datacube.load` being a complex beast. Thanks @robbibt!

**Review by uchchwhash — APPROVED** [review]

**Comment by uchchwhash** [comment]
Oh seems like mergify is not active here anymore. Feel free to merge whenever @robbibt!

**Comment by uchchwhash** [comment]
Since nobody else objected to it, merging. Thanks @robbibt!
